### DISCUSSION TITLE
how to handle new cardinality types in drivers

### DISCUSSION THREAD

**Discussion by nsidnev** [geldata/gel#2818] [Q&A] [answered]
Hi! I'am updating my Elixir driver to be compatible with `EdgeDB Beta 3`. When I added support for TLS certificates, I found that my tests that fetch objects are now broken :(

I went through the documentation for EdgeDB internals and found that there are now 2 new cardinality types: `AT_MOST_ONE` and `AT_LEAST_ONE`. In general, I understand what this types are supposed to mean, but I don't understand how to handle them in my driver.

I tried looking at other drivers and found that they handle it in different ways. Rust driver has a definition for all 5 types, driver for Go has 3 types, just like Python driver (`NO_RESULT`, `AT_MOST_ONE`, `MANY`).

Right now I'm a little stuck with this because I don't understand how I should handle the results or the description and query execution. In the end, now I have 2 issues with this:
1. When I send `ONE` query cardinality, as I did it for `EdgeDB Beta 2`, I get a `BinaryProtocolError` error response from `EdgeDB` server:
```
23:05:48.310 [error] EdgeDB.Connection (#PID<0.261.0>) disconnected: ** (EdgeDB.Protocol.Error) unknown expected cardinality "A"
```

2. If I send `AT_MOST_ONE`, as in Go, then I get a cardinality with type `ONE` in the output type descriptor in `CommandDataDescription ` from the server.

Could you please help me and explain what is happening here and what should I do to handle it correctly? At the moment I just added new cardinality types to my `Cardinality` module and handle result as if `cardinality == AT_MOST_ONE`, but I'm not sure if this is the right way.

**Comment by elprans** [comment, marked as answer]
`AT_MOST_ONE` means an expected result with 0 or 1 elements.  `CommandDataDescription` returns the actual inferred cardinality.  In this case `ONE`, which means that the query always returns a non-empty singleton set.  Cardinality `ONE` is a subset of `AT_MOST_ONE`, so it satisfies the client's assertion.

> Could you please help me and explain what is happening here and what should I do to handle it correctly?

Our bindings generally have two query methods: `query` and `query_single`.  The first does not make any sort of cardinality assertion, whereas the second expects the query to have the inferred result of `AT_MOST_ONE`.  You don't normally need to inspect the cardinality returned in `CommandDataDescription`, as it will always satisfy the requested cardinality (if the query does not satisfy it, an error is raised).

**Reply by nsidnev** [reply]
Hmm... Then it looks like the current way of handling the results in the driver is correct, thanks.

> You don't normally need to inspect the cardinality returned in CommandDataDescription, as it will always satisfy the requested cardinality (if the query does not satisfy it, an error is raised).

Yeap, that's exactly what the driver is doing right now. Perhaps I was not very clear about this in my question. I don't manually inspect the cardinality type, but just use the expected type provided by user's query to decide how execution result should be fetched (as a single value, an `EdgeDB.Set` structure instance, or just `:ok` status for scripts). I was just curious about this change and wanted to make sure I understand it correctly.

Except for this part, the migration was easy and went without problems. Thanks for new `EdgeDB` release! :)

**Reply by elprans** [reply]
Yeah, a change from `ONE` to `AT_MOST_ONE` was basically a fix, since we always meant the latter in this case.
